File: TinyScript/js/src/lexer/Lexer.js
const PeekIterator = require("../common/PeekIterator");
const Token = require("./Token");
const TokenType = require("./TokenType");
const AlphabetHelper = require("./AlphabetHelper");
const LexicalException = require("./LexicalException");
const arrayToGenerator = require("../common/arrayToGenerator");
const PeekTokenIterator = require("../parser/util/PeekTokenIterator");
const fs = require("fs");

class Lexer {
  analyse(source) {
    const tokens = [];
    const it = new PeekIterator(source, "\0");

    while (it.hasNext()) {
      let c = it.next();
      if (c == "\0") {
        break;
      }
      let lookahead = it.peek();

      if (c == " " || c == "\n" || c == "\r") {
        continue;
      }

      // 提取注释的程序
      if (c == "/") {
        if (lookahead == "/") {
          while (it.hasNext() && (c = it.next()) != "\n");
          continue;
        } else if (lookahead == "*") {
          let valid = false;
          while (it.hasNext()) {
            const p = it.next();
            if (p == "*" && it.peek() == "/") {
              valid = true;
              it.next();
              break;
            }
          }

          if (!valid) {
            throw new LexicalException("comment not matched");
          }
          continue;
        }
      }

      if (c == "{" || c == "}" || c == "(" || c == ")") {
        tokens.push(new Token(TokenType.BRACKET, c));
        continue;
      }

      if (c == '"' || c == "'") {
        it.putBack();
        tokens.push(Token.makeString(it));
        continue;
      }

      if (AlphabetHelper.isLetter(c)) {
        it.putBack();
        tokens.push(Token.makeVarOrKeyword(it));
        continue;
      }

      if (AlphabetHelper.isNumber(c)) {
        it.putBack();
        tokens.push(Token.makeNumber(it));
        continue;
      }

      // + -
      if ((c == "+" || c == "-") && AlphabetHelper.isNumber(lookahead)) {
        // 跳过:a+1, 1+1
        // +5, 3*-5
        const lastToken = tokens[tokens.length - 1] || null;

        if (lastToken == null || !lastToken.isValue()) {
          it.putBack();
          tokens.push(Token.makeNumber(it));
          continue;
        }
      }

      if (AlphabetHelper.isOperator(c)) {
        it.putBack();
        tokens.push(Token.makeOp(it));
        continue;
      }

      throw LexicalException.fromChar(c);
    }
    return tokens;
  }

  static fromFile(src) {
    const content = fs.readFileSync(src, "utf-8")
    const lexer = new Lexer()
    return arrayToGenerator(lexer.analyse(arrayToGenerator(content)))
  }
}

module.exports = Lexer;

File: TinyScript/java/src/main/java/lexer/Lexer.java
package lexer;

import common.AlphabetHelper;
import common.PeekIterator;

import java.io.*;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.stream.Stream;

public class Lexer {

    public ArrayList<Token> analyse(PeekIterator<Character> it) throws LexicalException {
        var tokens = new ArrayList<Token>();

        while(it.hasNext()) {
            char c = it.next();

            if(c == 0) {
                break;
            }
            char lookahead = it.peek();

            if(c == ' ' || c == '\n') {
                continue;
            }

            // 删除注释
            if(c == '/') {
                if(lookahead == '/') {
                    while(it.hasNext() && (c = it.next()) != '\n') {};
                    continue;
                }
                else if(lookahead == '*') {
                    it.next();//多读一个* 避免/*/通过
                    boolean valid = false;
                    while(it.hasNext()) {
                        char p = it.next();
                        if(p == '*' && it.peek() == '/') {
                            it.next();
                            valid = true;
                            break;
                        }
                    }
                    if(!valid) {
                        throw new LexicalException("comments not match");
                    }
                    continue;
                }
            }

            if(c == '{' || c == '}' || c == '(' || c == ')') {
                tokens.add(new Token(TokenType.BRACKET, c+""));
                continue;
            }

            if(c == '"' || c == '\'') {
                it.putBack();
                tokens.add(Token.makeString(it));
                continue;
            }

            if(AlphabetHelper.isLetter(c)) {
                it.putBack();;
                tokens.add(Token.makeVarOrKeyword(it));
                continue;
            }


            if(AlphabetHelper.isNumber(c)) {
                it.putBack();
                tokens.add(Token.makeNumber(it));
                continue;
            }


            if((c == '+' || c == '-' || c == '.') && AlphabetHelper.isNumber(lookahead)) {
                var lastToken = tokens.size() == 0 ? null : tokens.get(tokens.size() - 1);

                if(lastToken == null || !lastToken.isValue() || lastToken.isOperator()) {
                    it.putBack();
                    tokens.add(Token.makeNumber(it));
                    continue;
                }
            }

            if(AlphabetHelper.isOperator(c)) {
                it.putBack();
                tokens.add(Token.makeOp(it));
                continue;
            }

            throw new LexicalException(c);
        } // end while
        return tokens;
    }

   public ArrayList<Token> analyse(Stream source) throws LexicalException {
        var it = new PeekIterator<Character>(source, (char)0);
        return this.analyse(it);
    }

    /**
     * 从源代码文件加载并解析
     * @param src
     * @return
     * @throws FileNotFoundException
     * @throws UnsupportedEncodingException
     * @throws LexicalException
     */
    public static ArrayList<Token> fromFile(String src) throws FileNotFoundException, UnsupportedEncodingException, LexicalException {
        var file = new File(src);
        var fileStream = new FileInputStream(file);
        var inputStreamReader = new InputStreamReader(fileStream, "UTF-8");

        var br = new BufferedReader(inputStreamReader);


        /**
         * 利用BufferedReader每次读取一行
         */
        var it = new Iterator<Character>() {
            private String line = null;
            private int cursor = 0;

            private void readLine() throws IOException {
                if(line == null || cursor == line.length()) {
                    line = br.readLine();
                    cursor = 0;
                }
            }
            @Override
            public boolean hasNext() {
                try {
                    readLine();
                    return line != null;
               } catch (IOException e) {
                    return false;
                }
            }

            @Override
            public Character next() {
                try {
                    readLine();
                    return line != null ? line.charAt(cursor++) :null;
                } catch (IOException e) {
                    return null;
                }
            }
        };

        var peekIt = new PeekIterator<Character>(it, '\0');

        var lexer = new Lexer();
        return lexer.analyse(peekIt);

    }
}
